samples/qtask_ordered.c

```diff
@@ -0,0 +1,349 @@
+        }
+    }
+    gcratio = gc / (float) bamdata->core.l_qseq;
+
+    if (bam_aux_append(bamdata, "xr", 'f', sizeof(gcratio), (const uint8_t*)&gcratio) < 0) {
+        printf("Failed to add aux tag xr, errno: %d\n", errno);
+        ret = -1;
+    }
+    return ret;
+}
+
+
+/// getbamstorage - allocates storage for alignments to queue
+/** @param chunk no of alignments to be queued together
+returns allocated data.
+*/
+data* getbamstorage(int chunk)
+{
+    int i = 0;
+    data *bamdata = malloc(sizeof(data));
+    if (!bamdata) {
+        return NULL;
+    }
+    bamdata->bamarray = malloc(chunk * sizeof(bam1_t*));
+    if (!bamdata->bamarray) {
+        return NULL;
+    }
+    for (i = 0; i < chunk; ++i) {
+        bamdata->bamarray[i] = bam_init1();
+    }
+    bamdata->count = 0;
+    bamdata->size = chunk;
+
+    return bamdata;
+}
+
+/// thread_ordered_proc - does the processing of task in queue and queues the output back
+/** @param args pointer to set of data to be processed
+returns the processed data
+the processing could be in any order based on the number of threads in use but read of output
+from queue will be in order
+*/
+void *thread_ordered_proc(void *args)
+{
+    int i = 0;
+    data *bamdata = (data*)args;
+    for ( i = 0; i < bamdata->count; ++i) {
+        //add count
+        if (addcount(bamdata->bamarray[i]) < 0) {
+            printf("Failed to calculate gc data\n");
+            break;
+        }
+    }
+    return bamdata;
+}
+
+/// threadfn_orderedwrite - thread that read the output from queue and writes
+/** @param args pointer to data specific for the thread
+returns NULL
+*/
+void *threadfn_orderedwrite(void *args)
+{
+    orderedwrite *tdata = (orderedwrite*)args;
+    hts_tpool_result *r = NULL;
+    data *bamdata = NULL;
+    int i = 0;
+
+    struct timeval now;
+    struct timespec timeout;
+
+    pthread_mutex_lock(tdata->lock);                //lock to check the exit condition
+    do {
+        //get result and write; wait if no result is in queue - until shutdown of queue
+        while ((r = hts_tpool_next_result(tdata->queue))) {
+            bamdata = (data*) hts_tpool_result_data(r);
+            for (i = 0; i < bamdata->count; ++i) {
+                if (sam_write1(tdata->outfile, tdata->samhdr, bamdata->bamarray[i]) < 0) {
+                    printf("Failed to write output data\n");
+                    break;
+                }
+            }
+            hts_tpool_delete_result(r, 0);          //release the result memory
+            if (bamdata) {
+                for (i = 0; i < bamdata->size; ++i) {
+                    bam_destroy1(bamdata->bamarray[i]); //clear the bamdata;
+                }
+                free(bamdata->bamarray);
+                free(bamdata);
+            }
+        }
+        //no more data in queues, check and wait till exit is triggered
+        gettimeofday(&now, NULL);
+        timeout.tv_sec = now.tv_sec + WAIT;            //atmost 1 sec wait
```

Comment:
The possible usage of past time was missed. This may cause the thread to exit and thus leave the output queue full, stalling the process.
The loop condition needs update to continue on anything but exit trigger (0) to handle this past time value usage.
The working planned was as below.
Task queued by main thread, task done by worker threads and result enqueued to result queue. User thread polls for results and writes to file until exit set by main thread. At end of queueing main thread does queue flush - to finish all tasks - and waits for an interval before triggering exit. This interval is to ensure that the user thread makes result check at least once to save all results.